File: 6-using-regular-expressions/7.rb
# Normal backreferences
def format_date1(date)
  date_regex = /\A(\d{4})([\-\/])(\d{2})\2(\d{2})\z/
  p date.sub(date_regex, '\4.\3.\1')
end

format_date1('2016-06-17') # -> '17.06.2016'
format_date1('2017/05/03') # -> '03.05.2017'
format_date1('2015/01-31') # -> '2015/01-31' (no change)

# Named capture groups both inside and outside the regex
def format_date2(date)
  date_regex = /\A(?<year>\d{4})(?<delimiter>[\-\/])(?<month>\d{2})\k<delimiter>(?<day>\d{2})\z/
  p date.sub(date_regex, '\k<day>.\k<month>.\k<year>')
end  

format_date2('2016-06-17') # -> '17.06.2016'
format_date2('2017/05/03') # -> '03.05.2017'
format_date2('2015/01-31') # -> '2015/01-31' (no change)

# Named capture groups using symbol key to a MatchData object
def format_date3(date)
  date_regex = /\A(?<year>\d{4})(?<delimiter>[\-\/])(?<month>\d{2})\k<delimiter>(?<day>\d{2})\z/
  if match = date.match(date_regex)
    p "#{match[:day]}.#{match[:month]}.#{match[:year]}"
  else
    p date # This is needed to output the unmodified string when it fails, otherwise it would be nil.
  end
end

format_date3('2016-06-17') # -> '17.06.2016'
format_date3('2017/05/03') # -> '03.05.2017'
format_date3('2015/01-31') # -> '2015/01-31' (no change)

File: 6-using-regular-expressions/7.js
// Normal backreferences
const formatDate1 = function(date) {
  console.log(date.replace(/^(\d{4})([\-\/])(\d{2})\2(\d{2})$/, '$4.$3.$1'));
  return date.replace(/^(\d{4})([\-\/])(\d{2})\2(\d{2})$/, '$4.$3.$1');
}

formatDate1('2016-06-17'); // -> '17.06.2016'
formatDate1('2017/05/03'); // -> '03.05.2017'
formatDate1('2015/01-31'); // -> '2015/01-31' (no change)

/*
Named capture groups both inside and outside the regex
Note the \k<..> syntax for backreferencing a named capture group within a regex (same as Ruby)
Note the $<..> syntax for backreferencing a named capture group in a replacement string.
*/
const formatDate2 = function(date) {
  console.log(date.replace(/^(?<year>\d{4})(?<delimiter>[\-\/])(?<month>\d{2})\k<delimiter>(?<day>\d{2})$/, '$<day>.$<month>.$<year>'));
  return date.replace(/^(?<year>\d{4})(?<delimiter>[\-\/])(?<month>\d{2})\k<delimiter>(?<day>\d{2})$/, '$<day>.$<month>.$<year>');
}

formatDate2('2016-06-17'); // -> '17.06.2016'
formatDate2('2017/05/03'); // -> '03.05.2017'
formatDate2('2015/01-31'); // -> '2015/01-31' (no change)
